core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -542,9 +556,7 @@ assert equalType(
                     equivalents.remove(slots[i], equi.iterator().next());
                   }
                 }
-                assert result.result.rowType.equals(result.call.query.rowType)
-                    : Pair.of(result.result, result.call.query);
-                equivalents.put(result.result, result.call.query);
+                equivalents.put(result1, result.call.query);
```

Comment:
I think that we could keep the assert, but using `result1` instead of `result.result`:
```
assert result1.rowType.equals(result.call.query.rowType)
                    : Pair.of(result1, result.call.query);
```

Reply:
Hmm, this change only compensate the "top" target, but there's chance that child-node of target is equivalent with child-node of query, while they have different row-types. That's why I removed this this `assert`. If we keep it here, the test case I added will fail

Reply:
After second thinking, I decided to add back the `assert` for safe check, but the check only happens when `targetDescendant==target`

Reply:
Ok